src/main/java/net/imglib2/parallel/ForkJoinExecutorService.java

```diff
@@ -0,0 +1,146 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+package net.imglib2.parallel;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.ForkJoinTask;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * An ExecutorService that uses a ForkJoinPool to execute tasks,
+ * and therefor supports nested calls.
+ * <p>
+ * If used inside a {@link ForkJoinPool} that pool is used,
+ * in all other cases the common pool ({@link ForkJoinPool#commonPool()}) is used.
+ * <p>
+ * Shutdown and awaitTermination is not supported.
+ */
+public class ForkJoinExecutorService extends AbstractExecutorService
+{
+
+	public int getParallelism()
+	{
+		return getPool().getParallelism();
+	}
+
+	@Override
+	public void shutdown()
+	{
+	}
+
+	@Override
+	public List< Runnable > shutdownNow()
+	{
+		throw new UnsupportedOperationException( "ForkJoinExecutorService, shutdownNow is not implemented." );
+	}
+
+	@Override
+	public boolean isShutdown()
+	{
+		return false;
+	}
+
+	@Override
+	public boolean isTerminated()
+	{
+		return false;
+	}
+
+	@Override
+	public boolean awaitTermination( long l, TimeUnit timeUnit ) throws
+			InterruptedException
+	{
+		// NB: it's possible to implement this method. One might use a set of weak references to collect all tasks submitted.
+		// And this method call ForkJoinTask.get( long, timeUnit), to get the timing correct.
+		// But doing so introduces reduced performance, as the set of tasks needs to be managed.
+		// It's simpler to not use await termination at all.
+		// Alternative is to collect the futures and call get on them.
+		throw new UnsupportedOperationException( "ForkJoinExecutorService, awaitTermination is not implemented." );
+	}
+
+	@Override
+	public < T > List< Future< T > > invokeAll( Collection< ? extends Callable< T > > collection ) throws
+			InterruptedException
+	{
+		List< ForkJoinTask< T > >
+				futures = collection.stream().map( ForkJoinTask::adapt ).collect( Collectors
+				.toList() );
+		ForkJoinTask.invokeAll( futures );
+		return Collections.unmodifiableList( futures );
+	}
+
+	@Override
+	public < T > List< Future< T > > invokeAll( Collection< ? extends Callable< T > > collection, long l, TimeUnit timeUnit ) throws
+			InterruptedException
+	{
+		throw new UnsupportedOperationException( "ForkJoinExecutorService, invokeAll with timeout is not implemented." );
+	}
+
+	@Override
+	public Future< ? > submit( Runnable runnable )
+	{
+		return ForkJoinTask.adapt( runnable ).fork();
+	}
+
+	@Override
+	public < T > Future< T > submit( Runnable runnable, T t )
+	{
+		return ForkJoinTask.adapt( runnable, t ).fork();
+	}
+
+	@Override
+	public < T > Future< T > submit( Callable< T > callable )
+	{
+		return ForkJoinTask.adapt( callable ).fork();
```

Comment:
@maarzt Can you elaborate a bit on `ForkJoinExecutorService` design?

`ForkJoinPool` implements `ExecutorService`.
So this could also be implemented as
`return getPool().submit( callable )`.
And similar for the other methods.
Is there a reason to do it like this? What are the trade-offs?
I'm not getting the full picture here...

Reply:
We need `ForkJoinExecutorService` because it enables nested parallelization. Using `ForkJoinPool` for nested parallelization at some point throws an RejectedExecutionException: "Thread limit exceeded replacing blocked worker".

This happens, because `forkJoinPool.invokeAll(...)` is called inside a `ForkJoinThread`, which will block this thread until all the tasks are completed. The same is true for a call to `forkJoinPool.submit(...).get()`. The blocked thread can no longer be used for task execution. Which is bad as this wastes resources. `ForkJoinPool` mitigates this by creating new threads, that's what after some time causes the exception "Thread limit exceeded...".

So when you're inside a `ForkJoinPool` it's a bad idea to call `forkJoinPool.invokeAll` or `forkJoinPool.submit`. One should instead use `forkJoinTask.fork()` or `forkJoinTask.invokeAll()`.
And that's exactly what `ForkJoinExecutorService` does: it uses `ForkJoinTask.invokeAll(...)` instead of `forkJoinPool.invokeAll(...)`. And that makes `ForkJoinExecutorService` so much better for nested parallelization. 

And surprisingly `ForkJoinTask.invokeAll(...)` and `forkJoinTask.submit()` work anywhere, even outside a `ForkJoinPool`. That's why it's OK to use the `ForkJoinExecutorService` anywhere.

This advantages, have a downside too, `ForkJoinExecutorService` is not a fully functional `ExecutorService`, methods like `awaitTerminaten` or `shutdownNow` don't work.

Reply:
Now the differences between ForkJoinPool and ForkJoinExecutorService are explained in the javadoc.